### PR TITLE
Create CRUD endpoints for comments.

### PR DESCRIPTION
**Pull request by gurvirsasan**
Create the CRUD controller to:

- getComments
- getComment (by ID)
- Delete comment
- update comment
- create comment
 
connects up the controller methods to the routes.

### CODE REVIEW

File: server/routes/v1/index.ts [hunk 1]
```diff
@@ -1,10 +1,12 @@
 import express from 'express';
 import userRouter from './user';
 import postRouter from './post';
+import commentsRouter from './comment';
 
 const routes = express.Router();
 
 routes.use('/users/', userRouter);
 routes.use('/posts/', postRouter);
+routes.use('/comments/', commentsRouter);
```
Comment (andy9775) [inline server/routes/v1/index.ts:10]:
```suggestion
routes.use('/posts/:postid/comments/', commentsRouter);
```

You can now access the post the comment is being created for by `postid`

File: server/routes/v1/comment.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      req.body.body
+    );
+    res.status(result.status).json(result);
+  } catch (err) {
+    console.error(err);
+  }
+});
+
+commentRouter.put(
+  '/postid/comments/:commentid',
```
Comment (andy9775) [inline server/routes/v1/comment.ts:69]:
```suggestion
  '/:commentid',
```

File: server/routes/v1/comment.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      getAuthUser(res).id,
+      req.params.commentid
+    );
+    res.status(result.status).json(result);
+  } catch (err) {
+    console.error(err);
+  }
+});
+
+commentRouter.post(':postid/comments/', async (req: Request, res: Response) => {
```
Comment (andy9775) [inline server/routes/v1/comment.ts:55]:
```suggestion
commentRouter.post('', async (req: Request, res: Response) => {
```

File: server/controllers/v1/comment.ts [hunk 4]
```diff
@@ -0,0 +1,211 @@
+import { Comment } from '../../models/comment';
+import db from '../../models';
+
+// The return type of a Comment associated with the Post's User.
+export type CommentsUser = Comment & {
+  User: {
+    id: string;
+    firstName: string;
+    lastName: string;
+    createdAt: string;
```
Comment (andy9775) [inline server/controllers/v1/comment.ts:10]: No need for `createdAt` for the user data.
Here and elsewhere.

File: server/controllers/v1/tests/comment.test.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        author.id,
+        comment.id
+      );
+      expect(result.status).toBe(204);
+
+      const findResult = await commentController.getComment(comment.id);
+      expect(findResult.status).toBe(404);
+    });
+
+    it('should not delete a different comment', async () => {
```
Comment (andy9775) [inline server/controllers/v1/tests/comment.test.ts:118]: `should not delete a comment not authored by a user`?

File: server/controllers/v1/tests/comment.test.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        author.id,
+        comment.id,
+        newBody
+      );
+
+      expect(result.status).toBe(200);
+      expect(result.data!.result!.body).toBe(newBody);
+    });
+
+    it('should not update the comment for invalid user', async () => {
```
Comment (andy9775) [inline server/controllers/v1/tests/comment.test.ts:86]: `should only update a users comment`?

File: server/controllers/v1/tests/comment.test.ts [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+      const result = await commentController.createComment(
+        'This is a new comment!This is a new comment!!',
+        author.id,
+        undefined
+      );
+      expect(result.status).toBe(400);
+    });
+  });
+
+  describe('post mutation', () => {
```
Comment (andy9775) [inline server/controllers/v1/tests/comment.test.ts:69]: `comment mutation`

File: server/controllers/v1/tests/comment.test.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        expect(result.data.result![index].id).toBe(comments[index].id);
+        expect(result.data.result![index].User.firstName).toBe(
+          author.firstName
+        );
+      }
+      expect(result.data.total).toBe(3);
+    });
+
+    it('should return an error for a missing comment', async () => {
+      const result = await commentController.getComment('bingBong');
```
Comment (andy9775) [inline server/controllers/v1/tests/comment.test.ts:40]:
```suggestion
      const result = await commentController.getComment('123-123');
```

This follows the structure of a UID making it more obvious what it is

**Review by andy9775 — CHANGES_REQUESTED** [review]
Good job, some minor comments.

File: server/controllers/v1/tests/comment.test.ts [hunk 9]
```diff
@@ -25,19 +25,19 @@ describe('Test v1 - Comment Controller', () => {
...
-        expect(result.data.result![index].User.firstName).toBe(
-          author.firstName
-        );
-      }
+      expect(result.data.result!.map((x) => x.id).sort()).toEqual(
+        comments.map((x) => x.id).sort()
+      );
+      expect(
+        result.data.result!.map((x) => x.User.firstName === author.firstName)
+      ).toBeTruthy();
```
Comment (andy9775) [inline server/controllers/v1/tests/comment.test.ts:34]:
```suggestion
      expect(result.data.result!.map((c) => c.id).sort()).toEqual(
        comments.map((c) => c.id).sort()
      );
      expect(
        result.data.result!.map((c) => c.User.firstName === author.firstName)
      ).toBeTruthy();
```
`c` for comment, we're not in math lol

File: server/controllers/v1/tests/comment.test.ts [hunk 10]
```diff
@@ -83,7 +83,7 @@ describe('Test v1 - Comment Controller', () => {
       expect(result.data!.result!.body).toBe(newBody);
     });
 
-    it('should not update the comment for invalid user', async () => {
+    it("should only update the author's comment", async () => {
```
Comment (andy9775) [inline server/controllers/v1/tests/comment.test.ts:86]:
```suggestion
    it('should only update the author's comment', async () => {
```

Enable format on save in vs code :)

**Review by andy9775 — APPROVED** [review]
Some minor nits, otherwise LGTM!

File: server/controllers/v1/tests/comment.test.ts (hunk 10 above)
Comment (gurvirsasan) [reply server/controllers/v1/tests/comment.test.ts:86]: I have that enabled, but with `\'s` it automatically results to the double quotes

File: server/routes/v1/comment.ts [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+import db from '../../models';
+import CommentController from '../../controllers/v1/comment';
+import { getAuthUser } from '../../middleware/auth';
+
+const commentRouter = express.Router();
+const commentController = new CommentController(db.Post);
+
+commentRouter.get('', async (req: Request, res: Response) => {
+  try {
+    const limit = req.body.limit;
```
Comment (plausibly) [inline server/routes/v1/comment.ts:11]: These should be `req.query` for a GET request, we cannot send a body.

File: server/routes/v1/comment.ts [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+  } catch (err) {
+    console.error(err);
+  }
+});
+
+commentRouter.post('', async (req: Request, res: Response) => {
+  try {
+    const result = await commentController.createComment(
+      getAuthUser(res).id,
+      req.params.postid,
```
Comment (plausibly) [inline server/routes/v1/comment.ts:59]: shouldnt this be req.body?

**Review by plausibly — CHANGES_REQUESTED** [review]

File: server/controllers/v1/tests/comment.test.ts (hunk 10 above)
Comment (andy9775) [reply server/controllers/v1/tests/comment.test.ts]: NVM. I didn't realize that there was an apostrophe. Ignore me

File: server/routes/v1/comment.ts (hunk 11 above)
Comment (gurvirsasan) [reply server/routes/v1/comment.ts:11]: makes sense!

File: server/controllers/v1/comment.ts [hunk 13]
```diff
@@ -0,0 +1,218 @@
+import { Comment } from '../../models/comment';
+import db from '../../models';
+
+// The return type of a Comment associated with the Post's User.
+export type CommentsUser = Comment & {
+  User: {
```
Comment (plausibly) [inline server/controllers/v1/comment.ts:6]: Check in postman to make sure this type is correct. I remember we had a similar mistake for PostUser, it was returning `UserId` separately from User

File: server/controllers/v1/comment.ts [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+      this.commentsRepo.count({ where: { PostId: postID } }),
+    ]);
+
+    if (!data[0]) {
+      return {
+        status: 404,
+        data: { message: `Post ${postID} could not be found` },
+      };
+    }
+    if (!data[1]) {
```
Comment (plausibly) [inline server/controllers/v1/comment.ts:65]: Why not just combine this if statement with the same one on line 59 with an `or`? We are returning `404` for both, so just update the status message with a general error message

**Review by plausibly — APPROVED** [review]
LGTM. Just some minor comments

**Comment by plausibly** [comment]
Note: need to update getPosts to return number of comments in that post

File: server/controllers/v1/comment.ts [hunk 15]
```diff
@@ -11,7 +11,7 @@ export type CommentsUser = Comment & {
 };
 
 // The maximum number of results to return.
-const MAX_RESULTS = 50;
+const MAX_RESULTS = 10;
```
Comment (andy9775) [inline server/controllers/v1/comment.ts:14]: Only 10 posts per query?

**Review by andy9775 — APPROVED** [review]
One minor comment: Normally you wouldn't/shouldn't change existing migration files. You would create a new migration. Since we're dumping the DB each time, it's fine. Just an FYI.

Otherwise LGTM!

File: server/controllers/v1/comment.ts (hunk 15 above)
Comment (plausibly) [reply server/controllers/v1/comment.ts:14]: Yea, I think this is better than having lots of scrolling on the post page

File: server/models/tests/testHelpers.ts [hunk 16]
```diff
@@ -153,13 +152,24 @@ export async function makeValidComment(
...
+export async function getAllTags(): Promise<Tag[]> {
+  return await TagModel.findAll();
+}
+
+export async function getPostTags(
+  postId: string
+): Promise<(Tag & PostTag)[] | undefined> {
+  const post = await PostModel.findByPk(postId);
+
+  return await post?.getTags();
```
Comment (danielduong8) [inline server/models/tests/testHelpers.ts:171]: couldn't post be null if no postId is found? then you would call getTags() on null

**Review by danielduong8 — APPROVED** [review]
lgtm

File: server/models/tests/testHelpers.ts (hunk 16 above)
Comment (plausibly) [reply server/models/tests/testHelpers.ts:171]: its just for tests, but this looks like it was from the Tags PR.

**Pull request merged**